### PR TITLE
Use target features section in wasm-opt

### PR DESCRIPTION
**Pull request by tlively**
If the user does not supply features explicitly on the command line,
read and use the features in the target features section for
validation and passes. If the user does supply features explicitly,
error if they are not a superset of the features marked as used in the
target features section.

### CODE REVIEW

File: src/tools/tool-options.h [hunk 1]
```diff
@@ -33,88 +34,98 @@ struct ToolOptions : public Options {
...
+        .add("--no-validation", "-n",
+             "Disables validation, assumes inputs are correct",
              Options::Arguments::Zero,
              [this](Options* o, const std::string& argument) {
                passOptions.validate = false;
              });
-        ;
-  }
-
-  FeatureSet getFeatures() const {
```
Comment (kripken) [inline src/tools/tool-options.h]: why was this removed?

File: src/tools/tool-options.h [hunk 2]
```diff
@@ -25,6 +25,7 @@ namespace wasm {
 
 struct ToolOptions : public Options {
   PassOptions passOptions;
+  bool explicitFeatures = false;
```
Comment (kripken) [inline src/tools/tool-options.h]: this needs to be documented carefully, since I think it can be confusing.

File: src/wasm.h [hunk 3]
```diff
@@ -916,6 +916,8 @@ class Module {
 
   std::vector<UserSection> userSections;
   std::vector<std::string> debugInfoFileNames;
+  FeatureSet features = FeatureSet::MVP;
+  bool hasFeatures = false;
```
Comment (kripken) [inline src/wasm.h]: this worries me - we'll have features on both the module and the pass options. the latter feels like the only right place to me?

File: src/tools/tool-options.h (hunk 2 above)
Comment (kripken) [reply src/tools/tool-options.h]: this = the whole behavior of "mention one feature, need to use --autodetect-features to avoid an error, or mention them all".

File: src/tools/tool-options.h (hunk 1 above)
Comment (tlively) [reply src/tools/tool-options.h]: It didn't seem necessary since most other properties of the passOptions and related structs are accessed directly. I could alternatively have added a corresponding `setFeatures` method.

File: src/tools/tool-options.h (hunk 2 above)
Comment (tlively) [reply src/tools/tool-options.h]: Sounds good. Would option help strings be a reasonable place for this documentation or should it live out of band somewhere?

File: src/wasm.h (hunk 3 above)
Comment (tlively) [reply src/wasm.h]: I agree. I'll refactor to have `readFeatures` return the features directly to the caller instead of storing them on the module.

File: src/tools/tool-options.h (hunk 1 above)
Comment (kripken) [reply src/tools/tool-options.h]: Another option is to return the features by reference here, perhaps.

File: src/tools/tool-options.h (hunk 2 above)
Comment (kripken) [reply src/tools/tool-options.h]: Option help strings sound good I think.

**Comment by tlively** [comment]
With the new changes, the features are lazily calculated when `ToolOptions::getFeatures` is called. They cannot be eagerly calculated because the Module does not exist yet when options are parsed. Unfortunately because the passOptions inside ToolOptions is public, there is no way to enforce that the features are actually calculated before they are fetched out of the passOptions. The fix for this is to make PassOptions private and add accessor methods to ToolOptions that force the features to be computed before they are used.

File: src/wasm/wasm.cpp [hunk 4]
```diff
@@ -966,6 +973,71 @@ void Module::updateMaps() {
...
+    if (index + len > section->data.size()) {
+      // ill-formed string, silently fail
+      features = FeatureSet::MVP;
+      return false;
+    }
+    std::string name(&section->data[index], len);
+    index += len;
+
+    if (prefix == '-')
+      continue;
```
Comment (kripken) [inline src/wasm/wasm.cpp]: `{`, `}`

File: src/wasm/wasm.cpp [hunk 5]
```diff
@@ -966,6 +973,71 @@ void Module::updateMaps() {
...
+
+    if (name == BinaryConsts::UserSections::AtomicsFeature) {
+      features.setAtomics();
+    } else if (name == BinaryConsts::UserSections::BulkMemoryFeature) {
+      features.setBulkMemory();
+    } else if (name == BinaryConsts::UserSections::ExceptionHandlingFeature) {
+      // TODO: exception handling
```
Comment (kripken) [inline src/wasm/wasm.cpp]: maybe `WASM_UNREACHABLE()`, so we don't forget this later?

File: src/tools/wasm-opt.cpp [hunk 6]
```diff
@@ -253,7 +253,8 @@ int main(int argc, const char* argv[]) {
     auto runPasses = [&]() {
       options.runPasses(*curr);
       if (options.passOptions.validate) {
-        bool valid = WasmValidator().validate(*curr, options.getFeatures());
+        bool valid =
+            WasmValidator().validate(*curr, options.getFeatures(*curr));
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/wasm-opt.cpp]: Can be a PassOptions method to read the features, or something like that, and called soon after ModuleReader finishes?

[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/wasm-opt.cpp]: (in wasm-opt.cpp, and others can be similar)

File: src/tools/tool-options.h [hunk 7]
```diff
@@ -24,98 +24,108 @@
...
+        .addFeatureOpt(std::string("--disable-") + name, "",
+                       std::string("Disable ") + description,
+                       [=](FeatureSet& features) {
+                         features.set(feature, false);
+                       });
+    return *this;
+  }
+
+  FeatureSet& getFeatures(const Module& module) {
+    if (featuresInitialized)
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/tool-options.h]: ah, sorry, I got back into my LLVM style habit. I also caught myself giving local variables `UpperCase` names 😨

File: src/tools/wasm-opt.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/wasm-opt.cpp]: My rationale for doing it this way is that it makes it harder for users to forget to read the target features section (and if passOptions was private it would be impossible to forget). I also avoid the overhead of reading the target features section more than once by setting and checking `featuresInitialized`. If you still think the code would be more readable with a stateful API that requires users to remember to read the target features section as a separate step, I would be willing to reconsider.

[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/wasm-opt.cpp]: Good idea to focus on making it hard to forget reading the target features. One option might be to integrate this in `ModuleReader`. But I'm not sure if there would be cases where a wasm is loaded but we don't care about the features?

If we always care about the features, then that seems like an argument for putting the features on the Module object itself and not on PassOptions. That is, they are part of the Module, really; unlike other stuff on PassOptions like whether we ignore implicit traps - maybe?

[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/wasm-opt.cpp]: Yeah it's kind of weird because the module itself encodes a feature set, but not necessarily the one we want to use in passes. That's why I agree with you that it is not great for the module itself to expose the features. I think PassOptions is the best place to keep the features because they are meant to affect the behavior of the passes just like the rest of the pass options. I don't think there should be any cases where we need the PassOption's features but we don't have a module or don't want to consider the module's feature section.

File: src/tools/tool-options.h [hunk 8]
```diff
@@ -24,98 +24,109 @@
...
+             "Disables validation, assumes inputs are correct",
              Options::Arguments::Zero,
              [this](Options* o, const std::string& argument) {
                passOptions.validate = false;
              });
-        ;
   }
 
-  FeatureSet getFeatures() const {
+  ToolOptions &addFeatureOpt(const std::string& longName,
```
Comment (kripken) [inline src/tools/tool-options.h]: `&` on the type

File: src/tools/tool-options.h [hunk 9]
```diff
@@ -24,98 +24,109 @@
 namespace wasm {
 
 struct ToolOptions : public Options {
+  using Builder = std::function<void (FeatureSet&)>;
```
Comment (kripken) [inline src/tools/tool-options.h]: Perhaps `FeatureBuilder`? (we already have some other things called "Builder")

File: src/tools/tool-options.h [hunk 10]
```diff
@@ -24,98 +24,109 @@
...
+      }
+    } else if (wasmHasFeatures) {
+      passOptions.features = wasmFeatures;
+    }
+    featuresInitialized = true;
     return passOptions.features;
   }
+
+private:
+  std::vector<Builder> featureBuilders;
```
Comment (kripken) [inline src/tools/tool-options.h]: I have to say that a vector of lambdas scares me, because in C++ there isn't any type checking to make sure they didn't close on something that went away.

I don't understand why we need the lambdas, though - instead we can write user-requested features to say `userEnabledFeatures` and `userDisabledFeatures`, and then use that later in the computation?

File: src/tools/tool-options.h [hunk 11]
```diff
@@ -24,98 +24,109 @@
...
+                       })
+        .addFeatureOpt(std::string("--disable-") + name, "",
+                       std::string("Disable ") + description,
+                       [=](FeatureSet& features) {
+                         features.set(feature, false);
+                       });
+    return *this;
+  }
+
+  FeatureSet& getFeatures(const Module& module) {
```
Comment (kripken) [inline src/tools/tool-options.h]: As mentioned earlier I'd feel better about a `computeFeatures(Module&)` that does the computation and a `getFeatures()` that just immediately returns them.

File: src/tools/wasm-shell.cpp [hunk 12]
```diff
@@ -292,7 +292,7 @@ int main(int argc, const char* argv[]) {
         builders[moduleName].swap(builder);
         modules[moduleName].swap(module);
         i++;
-        bool valid = WasmValidator().validate(*modules[moduleName], options.getFeatures());
+        bool valid = WasmValidator().validate(*modules[moduleName], options.getFeatures(*modules[moduleName]));
```
Comment (kripken) [inline src/tools/wasm-shell.cpp]: with what I'm suggesting here, we'd replace changes like this to a tool like wasm-shell with a call to `computeFeatures(*modules[ModuleName])` at an earlier appropriate time

File: src/wasm.h [hunk 13]
```diff
@@ -954,6 +954,9 @@ class Module {
 
   void updateMaps();
 
+  // returns true if the module has a target features section
```
Comment (kripken) [inline src/wasm.h]: Perhaps another option is an `std::pair<FeatureSet, bool>` like the map etc. insert APIs? Or maybe just
```
// Read the features section into the out param, if it is present (if not, returns false).
bool readFeatureSection(FeatureSet&) const;
```
(i.e. change name + add commenting)
```

File: src/wasm.h [hunk 14]
```diff
@@ -954,6 +954,9 @@ class Module {
...
+  bool readFeatures(FeatureSet&) const;
+
```
Comment (kripken) [inline src/wasm.h]: I feel like this isn't the right place - we've kept `wasm.h` very minimal, leaving utilities to other places. Perhaps `wasm-binary.h` (since feature section is part of the binary format) or `ir/module-utils.h`?

File: src/tools/asm2wasm.cpp [hunk 15]
```diff
@@ -185,7 +185,7 @@ int main(int argc, const char *argv[]) {
   if (memInit != options.extra.end()) {
     if (options.runningDefaultOptimizationPasses()) {
       PassRunner runner(&wasm);
-      runner.setFeatures(options.getFeatures());
+      runner.setFeatures(options.getFeatures(wasm));
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/asm2wasm.cpp]: Sounds good.

File: src/tools/optimization-options.h [hunk 16]
```diff
@@ -130,7 +130,7 @@ struct OptimizationOptions : public ToolOptions {
   void runPasses(Module& wasm) {
     PassRunner passRunner(&wasm, passOptions);
     if (debug) passRunner.setDebug(true);
-    passRunner.setFeatures(passOptions.features);
+    passRunner.setFeatures(getFeatures(wasm));
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/optimization-options.h]: ok. I'll add an assert that the features are initialized to prevent users from accidentally using the default-initialized features.

**Comment by tlively** [comment]
@kripken I think you'll be happy with the latest changes here.
